### ISSUE TITLE
wrong port being sent to peers and sometimes to IRC

### ISSUE THREAD

**Issue by jcomeauictx** [jcomeauictx/AmericanCoin#3]
somehow the RPC port is being sent rather than the listen port, and sometimes I've seen IRC return the wrong port for a server.

```
/home/jcomeau/.americancoin/debug.log:receive version message: version 60001, blocks=361724, us=187.131.47.92:60410, them=108.161.139.100:9057, peer=108.161.139.100:9056
/home/jcomeau/.americancoin/debug.log:send version message: version 60001, blocks=361724, us=187.131.47.92:9057, them=108.161.139.100:9056, peer=108.161.139.100:9056
```

**Comment by jcomeauictx** [comment]
this could very well be the cause of issue #2 as well

```
IRC got who
Added 108.161.139.100:9057 from 62.210.131.147: 0 tried, 3 new
IRC got new address: 108.161.139.100:9057
connection timeout
```

**Comment by jcomeauictx** [comment]
and yet nothing from IRC ngrep shows port numbers, so it must be getting it from the americancoind software. ngrep output edited for readability
```
root@nutek:~# ngrep . host irc.smutfairy.com
interface: eth0 (108.161.139.0/255.255.255.0)
filter: (ip or ip6) and ( host irc.smutfairy.com )
match: .
#
T 62.210.131.147:6667 -> 108.161.139.100:47426 [AP]
  PING :irc.smutfairy.com..                                                  
##
T 108.161.139.100:47426 -> 62.210.131.147:6667 [AP]
  PONG :irc.smutfairy.com.                                                   
###
T 62.210.131.147:6667 -> 108.161.139.100:47426 [AP]
  ERROR :Closing Link: ns004.unternet.net (Remote host closed the connection)
  ..                                                                         
#######
T 62.210.131.147:6667 -> 108.161.139.100:49846 [AP]
  :irc.smutfairy.com NOTICE * :*** Looking up your hostname..                
##
T 62.210.131.147:6667 -> 108.161.139.100:49846 [AP]
  :irc.smutfairy.com NOTICE * :*** Found your hostname..                     
##
T 108.161.139.100:49846 -> 62.210.131.147:6667 [AP]
  NICK u76ynqpcY5RRjJx.                                                      
##
T 108.161.139.100:49846 -> 62.210.131.147:6667 [AP]
  USER u76ynqpcY5RRjJx 8 * : u76ynqpcY5RRjJx.                                
##
T 62.210.131.147:6667 -> 108.161.139.100:49846 [AP]
  :irc.smutfairy.com 001 u76ynqpcY5RRjJx :Welcome to the LFNet Internet Relay Chat Network u76ynqpcY5RRjJx..                                            
#
T 62.210.131.147:6667 -> 108.161.139.100:49846 [AP]
  :irc.smutfairy.com 002 u76ynqpcY5RRjJx :Your host is irc.smutfairy.com[::/6667], running version hybrid-8.2.8..
  :irc.smutfairy.com 003 u76ynqpcY5RRjJx 
  :This server was created Sep 11 2015 at 16:23:32..
  :irc.smutfairy.com 004 u76ynqpcY5RRjJx irc.smutfairy.com hybrid-8.2.8 DFGHRSWabcdefgijklnopqrsuwxy bchiklmnoprstveCIMORS bkloveIh..
  :irc.smutfairy.com 005 u76ynqpcY5RRjJx CALLERID CASEMAPPING=rfc1459 DEAF=D KICKLEN=180 MODES=6 PREFIX=(ohv)@%+ STATUSMSG=@%+ EXCEPTS INVEX NICKLEN=15 NETWORK=LFNet MAXLIST=beI:100 MAXTARGETS=4 :
  are supported by this server..
  :irc.smutfairy.com 005 u76ynqpcY5RRjJx CHANTYPES=# CHANLIMIT=#:25 CHANNELLEN=50 TOPICLEN=160 CHANMODES=beI,k,l,cimnprstC
  MORS ELIST=CMNTU SAFELIST WATCH=30 AWAYLEN=180 KNOCK :are supported by this
   server..
  :irc.smutfairy.com 042 u76ynqpcY5RRjJx 0LNAURCMW :your unique ID..
  :irc.smutfairy.com 251 u76ynqpcY5RRjJx :There are 0 users and 1880 invisible on 3 servers..
  :irc.smutfairy.com 252 u76ynqpcY5RRjJx 1 :IRC Operators online..
  :irc.smutfairy.com 254 u76ynqpcY5RRjJx 353 :channels formed..
  :irc.smutfairy.com 255 u76ynqpcY5RRjJx :I have 1682 clients and 1 servers..
  :irc.smutfairy.com 265 u76ynqpcY5RRjJx :Current local users: 1682  Max: 16318..
  :irc.smutfairy.com 266 u76ynqpcY5RRjJx :Current global users: 1880  Max: 17611..
  :irc.smutfairy.com 250 u76ynqpcY5RRjJx :Highest connection count: 16319 (16
  318 clients) (34388519 connections received)..
  :irc.smutfairy.com 375 u76ynqpcY5RRjJx :- irc.smutfairy.com Message of the Day - ..
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-   _      ______          _   ..
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-  | |    |  ____|        | |       _
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-  | |    | |__ _ __   ___| |_     | |
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-  | |    |  __| '_ \ / _ \ __|  . | |
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-  | |____| |  | | | |  __/ |_    \( )
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-  |______|_|  |_| |_|\___|\__|   (___)
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-                              ..
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-  LFNet listens on the following 
  ports:..
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-  Plaintext IRC: 6665, 6666, 6667, 6668, 6669, 7777..
  :irc.smutfairy.com 372 u76ynqpcY5RRjJx :-  TLS: 6679, 6697, 7000..
  :irc.smutfairy.com 376 u76ynqpcY5RRjJx :End of /MOTD comm
  and...:u76ynqpcY5RRjJx!u76ynqpcY5@ns004.unternet.net MODE u76ynqpcY5RRjJx :
  +i..                                                                       
##
T 108.161.139.100:49846 -> 62.210.131.147:6667 [AP]
  USERHOST u76ynqpcY5RRjJx.                                                  
#
T 62.210.131.147:6667 -> 108.161.139.100:49846 [AP]
  :irc.smutfairy.com 302 u76ynqpcY5RRjJx :u76ynqpcY5RRjJx=+u76ynqpcY5@108.161
  .139.100 ..                                                                
#
T 108.161.139.100:49846 -> 62.210.131.147:6667 [AP]
  NICK u76ynqpcYECrRVp.                                                      
#
T 62.210.131.147:6667 -> 108.161.139.100:49846 [AP]
  :u76ynqpcY5RRjJx!u76ynqpcY5@ns004.unternet.net NICK :u76ynqpcYECrRVp..     
#
T 108.161.139.100:49846 -> 62.210.131.147:6667 [AP]
  JOIN #americancoinCOM00.WHO #americancoinCOM00.                            
#
T 62.210.131.147:6667 -> 108.161.139.100:49846 [AP]
  :u76ynqpcYECrRVp!u76ynqpcY5@ns004.unternet.net JOIN :#americancoinCOM00..  
#
T 62.210.131.147:6667 -> 108.161.139.100:49846 [AP]
  :irc.smutfairy.com MODE #americancoinCOM00 +nt..
  :irc.smutfairy.com 353 u76ynqpcYECrRVp = #americancoinCOM00 :@u76ynqpcYECrRVp..
  :irc.smutfairy.com 366 u76ynqpcYECrRVp #americancoinCOM00 :End of /NAMES list...
  :irc.smutfairy.com 352 u76ynqpcYECrRVp #americancoinCOM00 u76ynqpcY5 ns004.unternet.net irc.smutfairy.com u76ynqpcYECrRVp H@ :0  u76ynqpcY5RRjJx..
  :irc.smutfairy.com 315 u76ynqpcYECrRVp #americancoinCOM00 :End of /WHO list...                   
#^Cexit
144 received, 0 dropped
```

**Comment by jcomeauictx** [comment]
irc.cpp is looking at the line ":u76ynqpcYECrRVp!u76ynqpcY5@ns004.unternet.net JOIN :#americancoinCOM00", taking "u76ynqpcYECrRVp" from the beginning up to the '!' into pszName and passing it to DecodeAddress

alternatively it takes the line ":irc.smutfairy.com 352 u76ynqpcYECrRVp #americancoinCOM00 u76ynqpcY5 ns004.unternet.net irc.smutfairy.com u76ynqpcYECrRVp H@ :0  u76ynqpcY5RRjJx" from the /WHO list and copies string index 7 into pszName

DecodeAddress treats the string past 'u' as a Base58-encoded bytestring, apparently IP number, port, and SHA256-derived checksum, although I can't match anything past the port number to the sha256sum. I get:

```
jcomeau@aspire:/usr/src/AmericanCoin/src$ echo 76ynqpcY5RRjJx | base58 -d | xxd -p
6ca18b642360a0c4b3f9
```
the first 4 digits match the IP address:

```
jcomeau@aspire:/usr/src/AmericanCoin/src$ printf "%x%x%x%x\n" 108 161 139 100
6ca18b64
```

and 0x2360 is 9056, not 9057:

```
jcomeau@aspire:/usr/src/AmericanCoin/src$ printf "%d\n" 0x2360
9056
```

**Comment by jcomeauictx** [comment]
DecodeAddress calls DecodeBase58Check in base58.h. see <https://people.xiph.org/~greg/irc.cpp> which seems a little clearer to me.

```
struct ircaddr
{
    struct in_addr ip;
    short port;
};
#pragma pack(pop)

string EncodeAddress(const CService& addr)
{
    struct ircaddr tmp;
    if (addr.GetInAddr(&tmp.ip))
    {
        tmp.port = htons(addr.GetPort());

        vector<unsigned char> vch(UBEGIN(tmp), UEND(tmp));
        return string("u") + EncodeBase58Check(vch);
    }
    return "";
}
```

**Comment by jcomeauictx** [comment]
waitaminute, I was using the wrong string:

```
jcomeau@aspire:/usr/src/AmericanCoin/src$ echo 76ynqpcYECrRVp | base58 -d | xxd -p
6ca18b642361f862509f
```
*that* gives 9057. I still can't figure the checksum though.

**Comment by jcomeauictx** [comment]
but it matches the GNU base58 utility:

```
jcomeau@aspire:/usr/src/AmericanCoin/src$ echo 6ca18b642361 | xxd -r -p | base58 -c; echo
76ynqpcYECrRVp
jcomeau@aspire:/usr/src/AmericanCoin/src$ echo 6ca18b642360 | xxd -r -p | base58 -c; echo
76ynqpcY5RRjJx
```

**Comment by jcomeauictx** [comment]
found out! it was sha256d, a double hash:

```
jcomeau@aspire:/usr/src/AmericanCoin/src$ echo 76ynqpcYECrRVp | base58 -d | xxd -p
6ca18b642361f862509f
jcomeau@aspire:/usr/src/AmericanCoin/src$ echo 76ynqpcY5RRjJx | base58 -d | xxd -p
6ca18b642360a0c4b3f9
jcomeau@aspire:/usr/src/AmericanCoin/src$ echo 6ca18b642361 | xxd -r -p | sha256sum | xxd -r -p | sha256sum
f862509fa35eaa2abf6b92c44eccd7c43b58d641ffd40fd55c3976f2a3397200  -
jcomeau@aspire:/usr/src/AmericanCoin/src$ echo 6ca18b642360 | xxd -r -p | sha256sum | xxd -r -p | sha256sum
a0c4b3f99a748573f67dc371eb6bc8c04b56f0445d0ecc4722f09082177b9987  -
```
now to figure out how the wrong port number is being sent to IRC in the first place.

**Comment by jcomeauictx** [comment]
irc.cpp twice encodes name used to NICK: `strMyName = EncodeAddress(GetLocalAddress(&addrConnect));`, at lines 257 and 292. this may account for the two different "names".

**Comment by jcomeauictx** [comment]
```
[IGNORE] :uBY22qb2nurDXZ4!x271475584@187.131.47.92 JOIN :#americancoinCOM00
WHO #americancoinCOM00
[IGNORE] :irc.smutfairy.com 352 irccat42 #americancoinCOM00 x271475584 187.131.47.92 irc.smutfairy.com uBY22qb2nurDXZ4 H :0  x271475584
[IGNORE] :irc.smutfairy.com 352 irccat42 #americancoinCOM00 irccat 187.131.47.92 irc.smutfairy.com irccat42 H :0  irccat
[IGNORE] :irc.smutfairy.com 352 irccat42 #americancoinCOM00 u76ynqpcY5 ns004.unternet.net irc.smutfairy.com u76ynqpcYECrRVp H@ :0  u76ynqpcY5RRjJx
[IGNORE] :irc.smutfairy.com 315 irccat42 #americancoinCOM00 :End of /WHO list.
```
current output at irc.lfnet.org (irc.smutfairy.com) while running latest code on my laptop. *it* is sending port 9056 (0x2360) while ns004.unternet.net is still showing port 9057.

jcomeau@aspire:~$ echo BY22qb2nurDXZ4 | base58 -d | xxd -p
bb832f5c236039a8f1ab

**Comment by jcomeauictx** [comment]
new code running on ns004, still fails, but now we can see it in real time:

```
GetLocal addr: 108.161.139.100:9056
GetLocal addr: 108.161.139.100:9056
IRC SENDING: NICK u76ynqpcY5RRjJx
IRC SENDING: USER u76ynqpcY5RRjJx 8 * : u76ynqpcY5RRjJx
IRC :irc.smutfairy.com 001 u76ynqpcY5RRjJx :Welcome to the LFNet Internet Relay Chat Network u76ynqpcY5RRjJx
IRC :irc.smutfairy.com 002 u76ynqpcY5RRjJx :Your host is irc.smutfairy.com[::/6667], running version hybrid-8.2.8
IRC :irc.smutfairy.com 003 u76ynqpcY5RRjJx :This server was created Sep 11 2015 at 16:23:32
IRC :irc.smutfairy.com 004 u76ynqpcY5RRjJx irc.smutfairy.com hybrid-8.2.8 DFGHRSWabcdefgijklnopqrsuwxy bchiklmnoprstveCIMORS bkloveIh
connection timeout
ERROR: GetMyExternalIP() : connection to 91.198.22.70:80 failed
GetMyExternalIP() received [] [::]:0
GetMyExternalIP() received [108.161.139.100] 108.161.139.100:0
GetMyExternalIP() returned 108.161.139.100
AddLocal(108.161.139.100:9056,5)
IRC SENDING: USERHOST u76ynqpcY5RRjJx
IRC :irc.smutfairy.com 302 u76ynqpcY5RRjJx :u76ynqpcY5RRjJx=+u76ynqpcY5@108.161.139.100 
GetIPFromIRC() got userhost 108.161.139.100
GetIPFromIRC() returned 108.161.139.100
AddLocal(108.161.139.100:9056,4)
GetLocal addr: 108.161.139.100:9057
IRC SENDING: NICK u76ynqpcYECrRVp
IRC SENDING: JOIN #americancoinCOM00
IRC SENDING: WHO #americancoinCOM00
IRC :u76ynqpcY5RRjJx!u76ynqpcY5@ns004.unternet.net NICK :u76ynqpcYECrRVp
connection timeout
IRC :u76ynqpcYECrRVp!u76ynqpcY5@ns004.unternet.net JOIN :#americancoinCOM00
IRC got join
Add of IRC address 108.161.139.100:9057 failed
IRC :irc.smutfairy.com MODE #americancoinCOM00 +nt
IRC :irc.smutfairy.com 353 u76ynqpcYECrRVp = #americancoinCOM00 :@u76ynqpcYECrRVp
IRC :irc.smutfairy.com 366 u76ynqpcYECrRVp #americancoinCOM00 :End of /NAMES list.
IRC :irc.smutfairy.com 352 u76ynqpcYECrRVp #americancoinCOM00 u76ynqpcY5 ns004.unternet.net irc.smutfairy.com u76ynqpcYECrRVp H@ :0  u76ynqpcY5RRjJx
IRC got who
Add of IRC address 108.161.139.100:9057 failed
IRC :irc.smutfairy.com 315 u76ynqpcYECrRVp #americancoinCOM00 :End of /WHO list.
ConnectNode(187.131.47.92:9056, (null), 0)
trying connection 187.131.47.92:9056 lastseen=2.0hrs
No valid UPnP IGDs found
connection timeout
ConnectNode(85.172.92.33:9056, (null), 0)
trying connection 85.172.92.33:9056 lastseen=19.8hrs
connection timeout
ConnectNode(187.131.47.92:9056, (null), 0)
trying connection 187.131.47.92:9056 lastseen=2.0hrs
connection timeout
```

maybe it has something to do with it having a "u"-name from the beginning, whereas my netbook starts off with an "x"-name.

**Comment by jcomeauictx** [comment]
latest commit has kludge to fix port, haven't dug enough yet to find out *why* the bad port is even making it into the address cache. however, my netbook is seeing the good address from IRC and *still* failing to connect: `Add of IRC address 108.161.139.100:9056 failed`

**Comment by jcomeauictx** [comment]
ns004 is no longer *sending* the wrong port, but it's still receiving the wrong port from my netbook running the same software. `grep 'us.*them' ~/.americancoin/debug.log` output:

```
send version message: version 60001, blocks=361724, us=108.161.139.100:9057, them=187.131.47.92:38744, peer=187.131.47.92:38744
receive version message: version 60001, blocks=0, us=108.161.139.100:9056, them=187.131.47.92:9057, peer=187.131.47.92:38744
send version message: version 60001, blocks=361724, us=108.161.139.100:9057, them=187.131.47.92:38836, peer=187.131.47.92:38836
receive version message: version 60001, blocks=0, us=108.161.139.100:9056, them=187.131.47.92:9057, peer=187.131.47.92:38836
send version message: version 60001, blocks=361724, us=0.0.0.0:0, them=0.0.0.0:0, peer=127.0.0.1:0
send version message: version 60001, blocks=361724, us=108.161.139.100:9056, them=187.131.47.92:39056, peer=187.131.47.92:39056
receive version message: version 60001, blocks=28500, us=108.161.139.100:9056, them=187.131.47.92:9057, peer=187.131.47.92:39056
```

**Comment by jcomeauictx** [comment]
verified fixed in 9a1b9c3. still don't know why that port is seen to begin with, but that's a problem for another day.

**Comment by jcomeauictx** [comment]
took another look at it tonight and spotted the bug right away. see line [246](https://github.com/jcomeauictx/AmericanCoin/blob/master/src/net.cpp#L246), where it *adds 1 to the port* in the mapLocalHost info corresponding to the address if the address is already stored. and on line [118](https://github.com/jcomeauictx/AmericanCoin/blob/master/src/net.cpp#L118) it's that incorrect port number being used by GetLocal.

so now I knew the problem, but had to wonder what the hell was the programmer thinking? so I decided to take a look at the Bitcoin sources and see if the bug was still there. [nope](https://github.com/bitcoin/bitcoin/blob/master/src/net.cpp#L227), it was intended to increment nScore, not nPort! which makes perfect sense! so I'm going to fix this shortly and remove my kludge.

**Comment by jcomeauictx** [comment]
fixed in 4b557e2
